adapters/info.go

```diff
@@ -1,13 +1,130 @@
 package adapters
 
 import (
+	"fmt"
 	"io/ioutil"
 
 	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
 	"github.com/prebid/prebid-server/openrtb_ext"
 	yaml "gopkg.in/yaml.v2"
 )
 
+// EnforceBidderInfo decorates the input Bidder by making sure that all the requests
+// to it are in sync with its static/bidder-info/{bidder}.yaml file.
+//
+// It adjusts incoming requests in the following ways:
+//   1. If App or Site traffic is not supported by the info file, then requests from
+//      those sources will be rejected before the delegate is called.
+//   2. If a given MediaType is not supported for the platform, then it will be set
+//      to nil before the request is forwarded to the delegate.
+//   3. Any Imps which have no MediaTypes left will be removed.
+//   4. If there are no valid Imps left, the delegate won't be called at all.
+func EnforceBidderInfo(bidder Bidder, info BidderInfo) Bidder {
+	return &InfoAwareBidder{
+		Bidder: bidder,
+		info:   info,
+	}
+}
+
+type InfoAwareBidder struct {
+	Bidder
+	info BidderInfo
+}
+
+func (i *InfoAwareBidder) MakeRequests(request *openrtb.BidRequest) ([]*RequestData, []error) {
+	var allowedMediaTypes []openrtb_ext.BidType
+	if request.Site != nil {
+		if i.info.Capabilities.Site == nil {
+			return nil, []error{BadInput("this bidder does not support site requests")}
+		}
+		allowedMediaTypes = i.info.Capabilities.Site.MediaTypes
+	}
+	if request.App != nil {
+		if i.info.Capabilities.App == nil {
+			return nil, []error{BadInput("this bidder does not support app requests")}
+		}
+		allowedMediaTypes = i.info.Capabilities.App.MediaTypes
+	}
+
+	// Filtering imps is quite expensive (array filter with large, non-pointer elements)... but should be rare,
+	// because it only happens if the publisher makes a really bad request.
+	//
+	// To avoid allocating new arrays and copying in the normal case, we'll make one pass to
+	// see if any imps need to be removed, and another to do the removing if necessary.
+	numToFilter, errs := i.pruneImps(request.Imp, allowedMediaTypes)
+	if numToFilter != 0 {
+		filteredImps, newErrs := i.filterImps(request.Imp, numToFilter)
+		request.Imp = filteredImps
+		errs = append(errs, newErrs...)
+	}
+	reqs, delegateErrs := i.Bidder.MakeRequests(request)
+	return reqs, append(errs, delegateErrs...)
+}
+
+// pruneImps trims invalid media types from each imp, and returns true if any of the
+// Imps have _no_ valid Media Types left.
+func (i *InfoAwareBidder) pruneImps(imps []openrtb.Imp, allowedTypes []openrtb_ext.BidType) (int, []error) {
```

Comment:
Could this be made more efficient by returning a list of the indexes that need to be pruned, so the second stage can loop over that rather than loop over the entire list of imps and recheck each one?

Reply:
I did think about that... but my hunch was that that would actually be less efficient. These lists are short, `malloc` is slow, and it's _probably_ not a huge deal because this code only runs if publishers send bad requests--which should hopefully only be done during testing and fixed before deploying to the internet.

I couldn't really tell you for sure which is more efficient without writing a benchmark for it, though. I can do that if you think it's needed.

Reply:
The malloc is a good point, I had not considered that. It probably will not be all that rare as lazy publishers will probably activate all bidders on all Imps as long as there is a single imp that each bidder can bid on.